src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -834,43 +836,73 @@ internal unsafe int GetNonRandomizedHashCode()
             }
         }
 
-        // Use this if and only if 'Denial of Service' attacks are not a concern (i.e. never used for free-form user input),
-        // or are otherwise mitigated
-        internal unsafe int GetNonRandomizedHashCodeOrdinalIgnoreCase()
+        internal int GetNonRandomizedHashCodeOrdinalIgnoreCase()
         {
-            fixed (char* src = &_firstChar)
-            {
-                Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
-                Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
+            return GetNonRandomizedHashCodeOrdinalIgnoreCaseStatic(ref _firstChar, Length, true);
+        }
 
-                uint hash1 = (5381 << 16) + 5381;
-                uint hash2 = hash1;
+        private static int GetNonRandomizedHashCodeOrdinalIgnoreCaseStatic(ref char firstChar, int length, bool normalizeNonAscii)
+        {
+            uint hash1 = (5381 << 16) + 5381;
+            uint hash2 = hash1;
 
-                uint* ptr = (uint*)src;
-                int length = this.Length;
+            // We "normalize to lowercase" every char by ORing with 0x0020. This casts
+            // a very wide net because it will change, e.g., '^' to '~'. But that should
+            // be ok because we expect this to be very rare in practice.
 
-                // We "normalize to lowercase" every char by ORing with 0x0020. This casts
-                // a very wide net because it will change, e.g., '^' to '~'. But that should
-                // be ok because we expect this to be very rare in practice.
+            const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
 
-                const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
+            int i = 0;
+            int count = 0;
 
-                while (length > 2)
+            while (count > 2)
```

Comment:
This loop will never execute?

Also beware: the original code assumed that it was working over _string_ instances, so it intentionally read past the end of the string to consume the null terminator. The new code might be working over stackalloc, or a char array, or some other buffer that's not guaranteed to be null terminated. Consuming that extra char could have unpredictable results.

Reply:
Oops, copied the wrong code, fixed (also, added `\0`). Now testing